src/EditorFeatures/TestUtilities/LanguageServer/AbstractLanguageServerProtocolTests.cs

```diff
@@ -452,16 +444,27 @@ private static LSP.DidCloseTextDocumentParams CreateDidCloseTextDocumentParams(U
         public sealed class TestLspServer : IDisposable
         {
             public readonly TestWorkspace TestWorkspace;
-            private readonly RequestDispatcher _requestDispatcher;
-            private readonly RequestExecutionQueue _executionQueue;
             private readonly Dictionary<string, IList<LSP.Location>> _locations;
+            private readonly VisualStudioInProcLanguageServer _languageServer;
+            private readonly JsonRpc _clientRpc;
+
+            public LSP.ClientCapabilities ClientCapabilities { get; }
 
-            internal TestLspServer(TestWorkspace testWorkspace)
+            private TestLspServer(TestWorkspace testWorkspace, LSP.ClientCapabilities clientCapabilities)
             {
                 TestWorkspace = testWorkspace;
+                ClientCapabilities = clientCapabilities;
                 _locations = GetAnnotatedLocations(testWorkspace, testWorkspace.CurrentSolution);
-                _requestDispatcher = CreateRequestDispatcher(testWorkspace);
-                _executionQueue = CreateRequestQueue(testWorkspace);
+
+                var (clientStream, serverStream) = FullDuplexStream.CreatePair();
```

Comment:
main change to switch to actual jsonrpc requests instead of directly adding messages to the request queue starts here